include/yaramod/types/regexp.h

```diff
@@ -23,41 +26,80 @@ namespace yaramod {
 class RegexpUnit
 {
 public:
-	RegexpUnit() {}
+	RegexpUnit()
+		: _tokenStream(std::make_shared<TokenStream>())
+	{
+	}
 	virtual ~RegexpUnit() {}
 
 	virtual std::string getText() const = 0;
 
 	virtual RegexpVisitResult accept(RegexpVisitor* v) = 0;
+	std::shared_ptr<TokenStream>&& getTokenStream() { return std::move(_tokenStream); }
+
+protected:
+	std::shared_ptr<TokenStream> _tokenStream;
 };
 
 /**
  * Class representing class unit in regular expressions. Class units
- * are enclosed in [] and can be either positive or negative (class starts with symbol ^).
+ * are enclosed in [] and can be either positive or negative (class starts with symbol ^, e.g. [^xyz]).
  */
 class RegexpClass : public RegexpUnit
 {
 public:
-	RegexpClass(std::string&& characters, bool negative = false) : _characters(characters), _negative(negative) {}
+	RegexpClass(const std::string& characters, bool negative = false)
+	{
+		_leftRectBracket = _tokenStream->emplace_back(TokenType::LSQB, "[");
+		_negative = _tokenStream->emplace_back(TokenType::REGEXP_CLASS_NEGATIVE, negative, negative? "^" : "");
+		addCharacters(std::move(characters));
+		_rightRectBracket = _tokenStream->emplace_back(TokenType::RSQB, "]");
+	}
 
-	virtual std::string getText() const
+	virtual std::string getText() const override
 	{
-		return '[' + (_negative ? "^" : std::string()) + _characters + ']';
+		std::ostringstream ss;
+		ss << _leftRectBracket->getPureText();
+		ss << _negative->getPureText();
+		for (TokenIt it : _characters)
+			ss << it->getPureText();
+		ss << _rightRectBracket->getPureText();
+
+		return ss.str();
 	}
 
 	virtual RegexpVisitResult accept(RegexpVisitor* v) override
 	{
 		return v->visit(this);
 	}
 
-	const std::string& getCharacters() const { return _characters; }
-	void setCharacters(const std::string& characters) { _characters = characters; }
+	void addCharacters(const std::string& text)
+	{
+		for (char c : text)
+			_characters.push_back(_tokenStream->emplace_back(TokenType::REGEXP_CHAR, std::string() + c));
+	}
+
+	std::string getCharacters() const
+	{
+		std::stringstream ss;
```

Comment:
I feel like these `stringstream` operations will hold us back on performance. `stringstream` is extremely slow compare to for example [libfmt](https://github.com/fmtlib/fmt) which already comes bundled with `pog`. It would probably make sense to do performance testing one more time so we can see whether there are possibly some issues with string stream performance.

Reply:
I have done performance testing on our cpp test suite builded with `RelWithDebInfo` build type. I searched for `stringstream` in the output in KCacheGrind and as you can see in the enclosed screenshot, there is not much time spent on `stringstream` operations.
![Screenshot from 2019-11-15 14-38-13](https://user-images.githubusercontent.com/29119229/68947420-b4156f00-07b5-11ea-93a7-7e7ca95a97f4.png)